### PR TITLE
Apply Cholesky decomposition for s.p.d. mapping matrices

### PR DESCRIPTION
**Pull request by DavidSCN**
## Main changes of this PR
Replaces the Eigen based QR decomposition by a Cholesky decomposition for s.p.d. matrices.

## Motivation and additional information
The current implementation requires C++17 in order to perform the decomposition immediately on the temporary matrix. There are other ways to handle this, but I don't see any particular reason not to use C++17 as we decided to upgrade anyway.

## Author's checklist

* [x] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I ran `make format` to ensure everything is formatted correctly.
* [x] I sticked to C++17 features.
* [x] I sticked to CMake version 3.16.3.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [x] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?

### CODE REVIEW

**Comment by IshaanDesai** [comment]
I compared the changes in this PR with v2.4.0 and ran tests for a polar cartesian mapping and the change in mapping compute time is an improvement of the order of approximately 10 to 13 times. The mapping compute times for various mesh sizes are:

Cartesian to polar mapping:

| number of vertices on read mesh | number of vertices on write mesh | mapping time for v2.4.0 (ms) | mapping time for this PR (ms) |
| -- | -- | -- | -- |
| 81 | 254 | 29 | 5 |
| 160 | 489 | 102 | 17 |
| 317 | 957 | 515 | 58 |
| 631 | 1885 | 3448 | 346 |
| 1259 | 2396 | 30588 | 2346 |
| 2516 | 7540 | 224210 | 16954 |


Polar to cartesian mapping:

| number of vertices on read mesh | number of vertices on write mesh | mapping time for v2.4.0 (ms) | mapping time for this PR (ms) |
| -- | -- | -- | -- |
| 144 | 162 | 18 | 3 |
| 296 | 320 | 58 | 6 |
| 596 | 634 | 258 | 26 |
| 1196 | 1262 | 1311 | 135 |
| 3777 | 2518 | 9837 | 820 |
| 4796 | 5032 | 73138 | 5593 |

With this change the mapping error computed in the l2 and infinity norm increased by ~1%

**Comment by DavidSCN** [comment]
Thanks a lot @IshaanDesai. The measurements are now a summary of #1320 #1335 #1338. I also ran small testcases on my local machine and observe a similar behavior. It's noteworthy that the RBF mapping is now also more robust (note the last two missing measurements for v2.4). First of all, the separated polynomial is more robust and second, the Cholesky decomposition is more robust, probably as it exploits the symmetry. A bit sad that the last changes here do not affect the thin-plate-splines, the volume-splines and the multi-quadrics, as the thin-plate-splines are in many places our default setting.
image
image

**Comment by fsimonis** [comment]
It would be fascinating to see how these changes affect the runtime of parallel cases in comparison to the PETSc implementation.

File: src/mapping/RadialBasisFctSolver.hpp [hunk 1]
```diff
@@ -204,5 +216,88 @@ RadialBasisFctSolver::RadialBasisFctSolver(RADIAL_BASIS_FUNCTION_T basisFunction
...
+  if (polynomial == Polynomial::SEPARATE) {
+    Eigen::VectorXd epsilon = _matrixV.transpose() * inputData;
+    PRECICE_ASSERT(epsilon.size() == _matrixV.cols());
+
+    // epsilon = Q^T * mu - epsilon (tau in the PETSc impl)
+    epsilon -= _matrixQ.transpose() * out;
+    PRECICE_ASSERT(epsilon.size() == _matrixQ.cols());
+
+    // out  = out - solveTranspose tau (sigma in the PETSc impl)
+#if EIGEN_VERSION_AT_LEAST(3, 4, 0)
```
Comment (fsimonis) [inline src/mapping/RadialBasisFctSolver.hpp:241]: Can you comment on what feature was introduced in 3.4.0 that required so much extra work before?

File: src/mapping/RadialBasisFctSolver.hpp [hunk 2]
```diff
@@ -204,5 +216,88 @@ RadialBasisFctSolver::RadialBasisFctSolver(RADIAL_BASIS_FUNCTION_T basisFunction
...
+    }
+#endif
+  }
+  return out;
+}
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+Eigen::VectorXd RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T>::solveConsistent(Eigen::VectorXd &inputData, Polynomial polynomial) const
+{
+  Eigen::VectorXd res;
```
Comment (fsimonis) [inline src/mapping/RadialBasisFctSolver.hpp:269]: Could we use a more expressive name please? :wink:

**Review by fsimonis — COMMENTED** [review]
The feeling I am getting in this PR is that we should probably lift Polynomial into the type system in the future.
This would simplify some places and move quite some run-time branches to compile-time decisions. It wouldn't surprise me if the compiler would consequently generate more efficient code.

**Comment by DavidSCN** [comment]
> The feeling I am getting in this PR is that we should probably lift Polynomial into the type system in the future.
This would simplify some places and move quite some run-time branches to compile-time decisions. It wouldn't surprise me if the compiler would consequently generate more efficient code.

I see your point. There are some places in the online phase, which could save some temporary allocations (depending on the branch). Do you want me to address this in this PR? It would definitely require some smart template instantiation selector mechanism, but I could look for a solution.

File: src/mapping/RadialBasisFctSolver.hpp [hunk 3]
```diff
@@ -169,15 +172,24 @@ Eigen::MatrixXd buildMatrixA(RADIAL_BASIS_FUNCTION_T basisFunction, const mesh::
 }
 
 template <typename RADIAL_BASIS_FUNCTION_T>
-RadialBasisFctSolver::RadialBasisFctSolver(RADIAL_BASIS_FUNCTION_T basisFunction, const mesh::Mesh &inputMesh, const mesh::Mesh &outputMesh, std::vector<bool> deadAxis, Polynomial polynomial)
+RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T>::RadialBasisFctSolver(RADIAL_BASIS_FUNCTION_T basisFunction, const mesh::Mesh &inputMesh, const mesh::Mesh &outputMesh, std::vector<bool> deadAxis, Polynomial polynomial)
 {
+  PRECICE_CHECK(!(RADIAL_BASIS_FUNCTION_T::isStrictlyPositiveDefinite() && polynomial == Polynomial::ON), "The integrated polynomial (polynomial=\"on\") is not supported for the selected radial-basis function. Please select another radial-basis function or change the polynomial configuration.");
```
[the remark this answers is not in the document]
Comment (DavidSCN) [reply src/mapping/RadialBasisFctSolver.hpp:177]: Not really, as the basis function class is passed as a template parameter and constructor argument to the mapping class. However, I can lift this check into the mapping class constructor and transform this check into an assert, which makes sense.

**Comment by fsimonis** [comment]
> Do you want me to address this in this PR?

No. It's a good next step in our iterative development process :wink:

File: tests/serial/mapping-rbf-gaussian/GaussianSupportRadius.xml [hunk 4]
```diff
@@ -21,7 +21,7 @@
       <use-mesh name="MeshTwo" provide="on" />
       <mapping:rbf-gaussian
         solver-rtol="1e-11"
-        polynomial="on"
+        polynomial="separate"
```
[the remark this answers is not in the document]
Comment (DavidSCN) [reply tests/serial/mapping-rbf-gaussian/GaussianSupportRadius.xml:24]: Integrated will no longer be supported after this PR is merged. The test here is only for the configuration of the Gaussian RBF for shape-parameter vs support-radius.

File: src/mapping/RadialBasisFctSolver.hpp (hunk 2 above)
Comment (DavidSCN) [reply src/mapping/RadialBasisFctSolver.hpp:269]: Yes

File: tests/serial/mapping-rbf-gaussian/GaussianSupportRadius.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/mapping-rbf-gaussian/GaussianSupportRadius.xml:24]: So, these changes are breaking, right?

[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/mapping-rbf-gaussian/GaussianSupportRadius.xml:24]: Nevermind, just read the check again.

**Review by fsimonis — APPROVED** [review]

File: src/mapping/RadialBasisFctSolver.hpp [hunk 5]
```diff
@@ -169,15 +172,24 @@ Eigen::MatrixXd buildMatrixA(RADIAL_BASIS_FUNCTION_T basisFunction, const mesh::
...
+RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T>::RadialBasisFctSolver(RADIAL_BASIS_FUNCTION_T basisFunction, const mesh::Mesh &inputMesh, const mesh::Mesh &outputMesh, std::vector<bool> deadAxis, Polynomial polynomial)
 {
+  PRECICE_ASSERT(!(RADIAL_BASIS_FUNCTION_T::isStrictlyPositiveDefinite() && polynomial == Polynomial::ON), "The integrated polynomial (polynomial=\"on\") is not supported for the selected radial-basis function. Please select another radial-basis function or change the polynomial configuration.");
   // Convert dead axis vector into an active axis array so that we can handle the reduction more easily
   std::array<bool, 3> activeAxis({{false, false, false}});
   std::transform(deadAxis.begin(), deadAxis.end(), activeAxis.begin(), [](const auto ax) { return !ax; });
+
   // First, assemble the interpolation matrix and check the invertability
-  _qrMatrixC = buildMatrixCLU(basisFunction, inputMesh, activeAxis, polynomial).colPivHouseholderQr();
+  bool decompositionConverged = false;
```
Comment (uekerman) [inline src/mapping/RadialBasisFctSolver.hpp:183]:
```suggestion
  bool decompositionSuccessful = false;
```
"converged" might be confusing for a decomposition (no iteration or similar).

**Review by uekerman — APPROVED** [review]
A small but very important step :+1:

I did not have a very thorough look as I guess that @fsimonis already did.
One minor renaming below.
 
We should change `use-qr-decomposition` to sth like `use-direct-solver` now right? Maybe easier if we do this in a second (breaking) PR. Could you please either open an issue or a PR?

**Comment by uekerman** [comment]
> It would be fascinating to see how these changes affect the runtime of parallel cases in comparison to the PETSc implementation.

and we should check CG again as alternative to our current one-fits-all GMRES.

**Pull request merged**